src/shared/acl-util.c

```diff
@@ -221,6 +221,13 @@ int parse_acl(const char *text, acl_t *acl_access, acl_t *acl_default, bool want
         STRV_FOREACH(entry, split) {
                 char *p;
 
+                p = endswith(*entry, "X");
+                if (p) {
+                        *acl_cond_exec = true;
+                        string_replace_char(*entry, 'X', 'x');
+                } else
+                        *acl_cond_exec = false;
```

Comment:
So this doesn't look right: you maintain a single variable here even if multiple entries are specified in the same ACL. so `u:foo:rwX,u:bar:rwx` would be treated as a lower case `x`, even though part of it uses the uppercase `X`.

I think it would be easiest to return two sets of ACLs: one for dir inodes and one for all other inode types, and build them up together.

i.e. have a prototype of:

```
int parse_acl(const char *text, acl_t *ret_acl_access, acl_t *ret_acl_access_dir, acl_t *ret_acl_default, bool want_mask);
```

Reply:
Your approach is mostly OK except for one problem: the execute bit should also be set when the file _already has execute permission for some user_ ([setfacl(1)](https://man.archlinux.org/man/setfacl.1.en#ACL_ENTRIES)). And other ways to take care of those? 🤔

Reply:
so, hmm, i guess that complicates stuff a bit.

So I think the ACL expression parser should probably return three `acl_t` lists:

1. the ACL to definitely apply to all inodes (as before)
2. the ACL to definitely apply as default ACL to directory inodes (as before)
3. and an ACL that contains rules that should be applied onto an inode which are subject to "X" mangling.

When applying an ACL to an inode we'd then take the 3rd set, iterate through it, and for each included entry set or unset the X bit depending on the existing x bit and the fact that it is a dir or not a dir. Each mangled entry is then added to a copy of the 1st set. The result is then the set actually applied to the inode.

It's a bit more messy, but I guess that's what it takes…